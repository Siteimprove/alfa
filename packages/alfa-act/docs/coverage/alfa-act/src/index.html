
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for alfa-act/src</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> alfa-act/src</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">0% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>0/513</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">0% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>0/242</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">0% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>0/433</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">0% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>0/380</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line low'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="audit.ts"><a href="audit.ts.html">audit.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="11" class="abs low">0/11</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="10" class="abs low">0/10</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="8" class="abs low">0/8</td>
	</tr>

<tr>
	<td class="file low" data-value="cache.ts"><a href="cache.ts.html">cache.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="8" class="abs low">0/8</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="7" class="abs low">0/7</td>
	</tr>

<tr>
	<td class="file low" data-value="diagnostic.ts"><a href="diagnostic.ts.html">diagnostic.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="12" class="abs low">0/12</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="8" class="abs low">0/8</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="16" class="abs low">0/16</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="11" class="abs low">0/11</td>
	</tr>

<tr>
	<td class="file empty" data-value="index.ts"><a href="index.ts.html">index.ts</a></td>
	<td data-value="0" class="pic empty">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

<tr>
	<td class="file low" data-value="interview.ts"><a href="interview.ts.html">interview.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="19" class="abs low">0/19</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="10" class="abs low">0/10</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="13" class="abs low">0/13</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="13" class="abs low">0/13</td>
	</tr>

<tr>
	<td class="file empty" data-value="oracle.ts"><a href="oracle.ts.html">oracle.ts</a></td>
	<td data-value="0" class="pic empty">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

<tr>
	<td class="file low" data-value="outcome.ts"><a href="outcome.ts.html">outcome.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="182" class="abs low">0/182</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="76" class="abs low">0/76</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="140" class="abs low">0/140</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="139" class="abs low">0/139</td>
	</tr>

<tr>
	<td class="file low" data-value="question.ts"><a href="question.ts.html">question.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="79" class="abs low">0/79</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="34" class="abs low">0/34</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="57" class="abs low">0/57</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="61" class="abs low">0/61</td>
	</tr>

<tr>
	<td class="file low" data-value="requirement.ts"><a href="requirement.ts.html">requirement.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="12" class="abs low">0/12</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="8" class="abs low">0/8</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="16" class="abs low">0/16</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="11" class="abs low">0/11</td>
	</tr>

<tr>
	<td class="file low" data-value="rule.ts"><a href="rule.ts.html">rule.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="183" class="abs low">0/183</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="88" class="abs low">0/88</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="165" class="abs low">0/165</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="124" class="abs low">0/124</td>
	</tr>

<tr>
	<td class="file low" data-value="tag.ts"><a href="tag.ts.html">tag.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="7" class="abs low">0/7</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="14" class="abs low">0/14</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="10" class="abs low">0/10</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	</tr>

<tr>
	<td class="file empty" data-value="tsconfig.json"><a href="tsconfig.json.html">tsconfig.json</a></td>
	<td data-value="0" class="pic empty">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-10-28T11:02:52.798Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    