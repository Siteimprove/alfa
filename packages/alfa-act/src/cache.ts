import { Future } from "@siteimprove/alfa-future";
import { Thunk } from "@siteimprove/alfa-thunk";

import { Outcome } from "./outcome";
import { Rule } from "./rule";

export class Cache {
  public static empty(): Cache {
    return new Cache();
  }

  private readonly _storage = new WeakMap<object, unknown>();

  private constructor() {}

  public get<I, T, Q>(
    rule: Rule<I, T, Q>,
    ifMissing: Thunk<Future<Iterable<Outcome<I, T, Q>>>>
  ): Future<Iterable<Outcome<I, T, Q>>> {
    let outcomes = this._storage.get(rule) as
      | Future<Iterable<Outcome<I, T, Q>>>
      | undefined;

    if (outcomes === undefined) {
      outcomes = ifMissing();
      this._storage.set(rule, outcomes);
    }

    return outcomes;
  }
}
