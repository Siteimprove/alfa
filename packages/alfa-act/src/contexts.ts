import { Context } from "@siteimprove/alfa-json-ld";

const Namespaces: Context = {
  /**
   * @see https://www.w3.org/TR/EARL/
   */
  earl: "http://www.w3.org/ns/earl#",

  /**
   * @see https://www.w3.org/TR/Content-in-RDF/
   */
  cnt: "http://www.w3.org/2011/content#",

  /**
   * @see https://www.w3.org/TR/HTTP-in-RDF/
   */
  http: "http://www.w3.org/2011/http#",

  /**
   * @see https://www.w3.org/TR/Pointers-in-RDF/
   */
  ptr: "https://www.w3.org/2009/pointers#",

  /**
   * @see https://www.w3.org/TR/WCAG/
   */
  wcag: "https://www.w3.org/TR/WCAG/#",

  /**
   * @see https://siteimprove.github.io/sanshikan/
   */
  sanshikan: "https://siteimprove.github.io/sanshikan/"
};

export namespace Contexts {
  /**
   * @see https://www.w3.org/TR/EARL/#Assertion
   */
  export const Assertion: Context = {
    ...Namespaces,

    // Properties

    /**
     * @see https://www.w3.org/TR/EARL/#assertedBy
     */
    assertedBy: {
      "@id": "earl:assertedBy",
      "@type": "earl:Assertor"
    },

    /**
     * @see https://www.w3.org/TR/EARL/#subject
     */
    subject: {
      "@id": "earl:subject",
      "@type": "earl:TestSubject",
      "@container": "@set"
    },

    /**
     * @see https://www.w3.org/TR/EARL/#test
     */
    test: {
      "@id": "earl:test",
      "@type": "earl:TestCriterion",
      "@container": "@set"
    },

    /**
     * @see https://www.w3.org/TR/EARL/#mode
     */
    mode: {
      "@id": "earl:mode",
      "@type": "earl:TestMode"
    },

    /**
     * @see https://www.w3.org/TR/EARL/#result
     */
    result: {
      "@id": "earl:result",
      "@type": "earl:TestResult",
      "@container": "@set"
    }
  };

  /**
   * @see https://www.w3.org/TR/EARL/#TestResult
   */
  export const Result: Context = {
    /**
     * @see https://www.w3.org/TR/EARL/#outcome
     */
    outcome: {
      "@id": "earl:outcome",
      "@type": "earl:OutcomeValue"
    },

    /**
     * @see https://www.w3.org/TR/EARL/#pointer
     */
    pointer: {
      "@id": "earl:pointer",
      "@type": "ptr:Pointer"
    },

    /**
     * @see https://www.w3.org/TR/EARL/#info
     */
    info: "earl:info"
  };

  /**
   * @see https://www.w3.org/TR/HTTP-in-RDF/#MessageClass
   */
  export const Message: Context = {
    ...Namespaces,

    // Properties

    /**
     * @see https://www.w3.org/TR/HTTP-in-RDF/#headersProperty
     */
    headers: {
      "@id": "http:headers",
      "@type": "http:MessageHeader",
      "@container": "@set"
    },

    /**
     * @see https://www.w3.org/TR/HTTP-in-RDF/#bodyProperty
     */
    body: {
      "@id": "http:body",
      "@type": "cnt:ContentAsText"
    }
  };

  /**
   * @see https://www.w3.org/TR/HTTP-in-RDF/#RequestClass
   */
  export const Request: Context = {
    ...Namespaces,
    ...Message,

    // Properties

    /**
     * @see https://www.w3.org/TR/HTTP-in-RDF/#methodNameProperty
     */
    methodName: "http:methodName",

    /**
     * @see https://www.w3.org/TR/HTTP-in-RDF/#requestURIProperty
     */
    requestURI: "http:requestURI"
  };

  /**
   * @see https://www.w3.org/TR/HTTP-in-RDF/#ResponseClass
   */
  export const Response: Context = {
    ...Namespaces,
    ...Message,

    // Properties

    /**
     * @see https://www.w3.org/TR/HTTP-in-RDF/#statusCodeValueProperty
     */
    statusCodeValue: "http:statusCodeValue"
  };

  /**
   * @see https://www.w3.org/TR/Content-in-RDF/#ContentClass
   */
  export const Content: Context = {
    ...Namespaces,

    // Properties

    /**
     * @see https://www.w3.org/TR/Content-in-RDF/#characterEncodingProperty
     */
    characterEncoding: "cnt:characterEncoding",

    /**
     * @see https://www.w3.org/TR/Content-in-RDF/#charsProperty
     */
    chars: "cnt:chars"
  };

  /**
   * @see https://www.w3.org/TR/Pointers-in-RDF/#offsetPointerClass
   */
  export const OffsetPointer: Context = {
    ...Namespaces,

    // Properties

    /**
     * @see https://www.w3.org/TR/Pointers-in-RDF/#referenceProperty
     */
    reference: {
      "@id": "ptr:reference",
      "@type": "earl:TestSubject"
    },

    /**
     * @see https://www.w3.org/TR/Pointers-in-RDF/#offsetProperty
     */
    offset: "ptr:offset"
  };

  /**
   * @see https://www.w3.org/TR/Pointers-in-RDF/#xPathPointerClass
   */
  export const XPathPointer: Context = {
    ...Namespaces,

    // Properties

    /**
     * @see https://www.w3.org/TR/Pointers-in-RDF/#referenceProperty
     */
    reference: {
      "@id": "ptr:reference",
      "@type": "earl:TestSubject"
    },

    /**
     * @see https://www.w3.org/TR/Pointers-in-RDF/#expressionProperty
     */
    expression: "ptr:expression"
  };
}
