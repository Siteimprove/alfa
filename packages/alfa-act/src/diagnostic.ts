import { Equatable } from "@siteimprove/alfa-equatable";
import { Serializable } from "@siteimprove/alfa-json";

import * as json from "@siteimprove/alfa-json";

export class Diagnostic implements Equatable, Serializable {
  public static of(message: string): Diagnostic {
    return new Diagnostic(normalize(message));
  }

  protected readonly _message: string;

  protected constructor(message: string) {
    this._message = message;
  }

  public get message(): string {
    return this._message;
  }

  public equals(value: unknown): value is this {
    return value instanceof Diagnostic && value._message === this._message;
  }

  public toJSON(): Diagnostic.JSON {
    return {
      message: this._message,
    };
  }
}

export namespace Diagnostic {
  export interface JSON {
    [key: string]: json.JSON;
    message: string;
  }

  export function isDiagnostic(value: unknown): value is Diagnostic {
    return value instanceof Diagnostic;
  }
}

function normalize(input: string): string {
  return input.trim().replace(/\s+/g, " ");
}
