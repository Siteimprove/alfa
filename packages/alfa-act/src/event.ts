import { Serializable } from "@siteimprove/alfa-json";

import * as json from "@siteimprove/alfa-json";

import { Rule } from "./rule";

/**
 * @public
 */
export class Event<
  INPUT,
  TARGET,
  QUESTION,
  SUBJECT,
  TYPE extends Event.Type = Event.Type,
  NAME extends string = string
> implements Serializable<Event.JSON<TYPE, NAME>>
{
  public static of<
    INPUT,
    TARGET,
    QUESTION,
    SUBJECT,
    TYPE extends Event.Type,
    NAME extends string
  >(
    type: TYPE,
    rule: Rule<INPUT, TARGET, QUESTION, SUBJECT>,
    name: NAME
  ): Event<INPUT, TARGET, QUESTION, SUBJECT, TYPE, NAME> {
    return new Event(type, rule, name);
  }

  private readonly _type: TYPE;
  private readonly _rule: Rule<INPUT, TARGET, QUESTION, SUBJECT>;
  private readonly _name: NAME;

  constructor(
    type: TYPE,
    rule: Rule<INPUT, TARGET, QUESTION, SUBJECT>,
    name: NAME
  ) {
    this._type = type;
    this._rule = rule;
    this._name = name;
  }

  public get type(): TYPE {
    return this._type;
  }

  public get rule(): Rule<INPUT, TARGET, QUESTION, SUBJECT> {
    return this._rule;
  }

  public get name(): NAME {
    return this._name;
  }

  public toJSON(): Event.JSON<TYPE, NAME> {
    return {
      type: this._type,
      rule: this._rule.toJSON(),
      name: this._name,
    };
  }
}

export namespace Event {
  export type Type = "start" | "end";

  export interface JSON<T extends Type = Type, N extends string = string> {
    [key: string]: json.JSON;
    type: T;
    rule: Rule.JSON;
    name: N;
  }

  export function isEvent<
    INPUT,
    TARGET,
    QUESTION,
    SUBJECT,
    TYPE extends Event.Type = Event.Type,
    NAME extends string = string
  >(
    value: unknown
  ): value is Event<INPUT, TARGET, QUESTION, SUBJECT, TYPE, NAME> {
    return value instanceof Event;
  }

  export function start<I, T, Q, S>(
    rule: Rule<I, T, Q, S>
  ): Event<I, T, Q, S, "start", "rule"> {
    return Event.of("start", rule, "rule");
  }

  export function end<I, T, Q, S>(
    rule: Rule<I, T, Q, S>
  ): Event<I, T, Q, S, "end", "rule"> {
    return Event.of("end", rule, "rule");
  }

  export function startApplicability<I, T, Q, S>(
    rule: Rule<I, T, Q, S>
  ): Event<I, T, Q, S, "start", "applicability"> {
    return Event.of("start", rule, "applicability");
  }

  export function endApplicability<I, T, Q, S>(
    rule: Rule<I, T, Q, S>
  ): Event<I, T, Q, S, "end", "applicability"> {
    return Event.of("end", rule, "applicability");
  }

  export function startExpectation<I, T, Q, S>(
    rule: Rule<I, T, Q, S>
  ): Event<I, T, Q, S, "start", "expectation"> {
    return Event.of("start", rule, "expectation");
  }

  export function endExpectation<I, T, Q, S>(
    rule: Rule<I, T, Q, S>
  ): Event<I, T, Q, S, "end", "expectation"> {
    return Event.of("end", rule, "expectation");
  }
}
