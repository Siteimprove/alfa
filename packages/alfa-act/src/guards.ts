import { Aspect, Atomic, Composite, Rule, Target } from "./types";

export function isAtomic<A extends Aspect, T extends Target>(
  rule: Rule<A, T>
): rule is Atomic.Rule<A, T> {
  return "composes" in rule === false;
}

export function isComposite<A extends Aspect, T extends Target>(
  rule: Rule<A, T>
): rule is Composite.Rule<A, T> {
  return "composes" in rule;
}
