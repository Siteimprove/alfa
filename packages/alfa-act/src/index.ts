export * from "./audit";
export * from "./guards";
export * from "./markdown";
export * from "./types";
