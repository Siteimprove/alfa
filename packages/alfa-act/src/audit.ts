import { Future } from "@siteimprove/alfa-future";
import { Iterable } from "@siteimprove/alfa-iterable";
import { List } from "@siteimprove/alfa-list";
import { None } from "@siteimprove/alfa-option";

import { Cache } from "./cache";
import { Oracle } from "./oracle";
import { Outcome } from "./outcome";
import { Rule } from "./rule";

export class Audit<I, T = unknown, Q = unknown> {
  public static of<I, T = unknown, Q = unknown>(
    input: I,
    oracle: Oracle<Q> = () => Future.now(None)
  ): Audit<I, T, Q> {
    return new Audit(input, oracle, List.empty());
  }

  private readonly _input: I;
  private readonly _oracle: Oracle<Q>;
  private readonly _rules: List<Rule<I, T, Q>>;

  private constructor(input: I, oracle: Oracle<Q>, rules: List<Rule<I, T, Q>>) {
    this._input = input;
    this._oracle = oracle;
    this._rules = rules;
  }

  public add(rule: Rule<I, T, Q>): Audit<I, T, Q> {
    return new Audit(this._input, this._oracle, this._rules.push(rule));
  }

  public evaluate(): Future<Iterable<Outcome<I, T, Q>>> {
    const outcomes = Cache.empty();

    return Future.traverse(this._rules, (rule) =>
      rule.evaluate(this._input, this._oracle, outcomes)
    ).map(Iterable.flatten);
  }
}
