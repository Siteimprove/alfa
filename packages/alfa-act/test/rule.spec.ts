import { describe, it } from "vitest";

import { Hash, Hashable } from "@siteimprove/alfa-hash";
import { Err, Ok } from "@siteimprove/alfa-result";

import { Diagnostic, Rule } from "../src";

class NumberWrapper implements Hashable {
  constructor(public value: number) {}
  hash(hash: Hash): void {
    hash.writeNumber(this.value);
  }
}

const dummyRule = Rule.Atomic.of<Array<NumberWrapper>, NumberWrapper>({
  uri: "positive numbers must be even",
  evaluate(numbers) {
    return {
      applicability() {
        return numbers.filter((n) => n.value >= 0);
      },

      expectations(target) {
        const result =
          target.value % 2 === 0
            ? Ok.of(Diagnostic.of("Number was even"))
            : Err.of(Diagnostic.of("Number was not even"));
        return {
          1: result,
        };
      },
    };
  },
});

describe("#evaluate()", () => {
  it("matches snapshot", async (ctx) => {
    const result = await dummyRule.evaluate([
      new NumberWrapper(-3),
      new NumberWrapper(-2),
      new NumberWrapper(0),
      new NumberWrapper(2),
      new NumberWrapper(3),
      new NumberWrapper(3.1415),
    ]);

    ctx.expect(result).toMatchSnapshot();
  });
});
