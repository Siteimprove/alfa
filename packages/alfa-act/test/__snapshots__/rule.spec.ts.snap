// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`#evaluate() > matches snapshot 1`] = `
[
  {
    "expectations": [
      [
        "1",
        {
          "type": "ok",
          "value": {
            "message": "Number was even",
          },
        },
      ],
    ],
    "mode": "automatic",
    "outcome": "passed",
    "rule": {
      "requirements": [],
      "tags": [],
      "type": "atomic",
      "uri": "positive numbers must be even",
    },
    "target": {
      "value": 0,
    },
  },
  {
    "expectations": [
      [
        "1",
        {
          "type": "ok",
          "value": {
            "message": "Number was even",
          },
        },
      ],
    ],
    "mode": "automatic",
    "outcome": "passed",
    "rule": {
      "requirements": [],
      "tags": [],
      "type": "atomic",
      "uri": "positive numbers must be even",
    },
    "target": {
      "value": 2,
    },
  },
  {
    "expectations": [
      [
        "1",
        {
          "error": {
            "message": "Number was not even",
          },
          "type": "err",
        },
      ],
    ],
    "mode": "automatic",
    "outcome": "failed",
    "rule": {
      "requirements": [],
      "tags": [],
      "type": "atomic",
      "uri": "positive numbers must be even",
    },
    "target": {
      "value": 3,
    },
  },
  {
    "expectations": [
      [
        "1",
        {
          "error": {
            "message": "Number was not even",
          },
          "type": "err",
        },
      ],
    ],
    "mode": "automatic",
    "outcome": "failed",
    "rule": {
      "requirements": [],
      "tags": [],
      "type": "atomic",
      "uri": "positive numbers must be even",
    },
    "target": {
      "value": 3.1415,
    },
  },
]
`;
